### PR TITLE
Feat/add transpiler, compress single qubit gates, and convert to phase shift and half x gates

### PR DESCRIPTION
**Pull request by fSosaRey**
…pression to universal gates

compare_circuits() and compare_kets() are implemented in the testing of the transpiler, but could be added to the library. They compare the output allowing for a possible global phase difference.

### CODE REVIEW

File: src/core/quantum_gate.jl [hunk 1]
```diff
@@ -1065,6 +1027,34 @@ get_gate_parameters(gate::Universal)=Dict(
     "lambda"=>gate.lambda
 )
 
+function get_universal(target::Integer,op::AbstractOperator)
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:1030]: Rename to `as_universal_gate`

File: src/core/transpile.jl [hunk 2]
```diff
@@ -1,13 +1,254 @@
...
+    end
+
+    ######################################################
+    #
+    #  boudaries is a Vector of Tuple{i_gate,targets}
+    #  where:
+    #       i_gate  : the gate's index in 'gates'
+    #       targets : the targets of gate[i_gate]
+    #               
+    boundaries=Vector{Tuple{Int,         Vector{Int}}}([])
```
Comment (plerouxanyon) [inline src/core/transpile.jl:96]: This spacing doesn't seem right

File: test/transpile.jl [hunk 3]
```diff
@@ -1,12 +1,236 @@
...
+    end
+end 
+
+@testset "Transpiler" begin
+    struct NonExistentTranspiler<:Transpiler end
+
+    @test_throws NotImplementedError transpile(NonExistentTranspiler(),QuantumCircuit(qubit_count = 2))
+end
+
+@testset "basic transpilation" begin
```
Comment (plerouxanyon) [inline test/transpile.jl:116]: Add the test that failed earlier

`q[1]:──X────Y─────────*──────────────*─────────Z──────────────*───────`
`                      |              |                        |      `
`q[2]:────────────H────|────X────Y────|────H───────────────────*───────`
`                      |              |                        |      `
`q[3]:─────────────────X──────────────|────────────────────────X───────`
`                                     |                               `
`q[4]:────────────────────────────────X──────────────X────Y─────────H──`

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: test/transpile.jl (hunk 3 above)
Comment (fSosaRey) [reply test/transpile.jl:116]: that circuit was tested in the next @test_set (defined as `input_gates`)

**Pull request merged**